pilot/pkg/proxy/envoy/v2/ads.go

```diff
@@ -705,17 +705,55 @@ func proxyNeedsPush(con *XdsConnection, targetNamespaces map[string]struct{}) bo
 	return false
 }
 
+// updateProxyServiceInstances determines whether the proxy associated with the given service instance
+// needs to refresh its local inbound listeners, and if so, triggers a push to that proxy.
+func (s *DiscoveryServer) updateProxyServiceInstances(instance *model.ServiceInstance) error {
+	instanceIP := instance.Endpoint.Address
+
+	// Check whether there's an existing connection from that IP address
+	adsClientsMutex.RLock()
+	con, ok := s.connectionsByIP[instanceIP]
+	adsClientsMutex.RUnlock()
+	if !ok {
+		return nil
+	}
+
+	proxy := con.modelNode
+	instances, err := s.Env.GetProxyServiceInstances(proxy)
+	if err != nil {
+		adsLog.Errorf("Error getting proxy %s service instances: %v", proxy.ID, err)
+		return err
+	}
+
+	if reflect.DeepEqual(instances, proxy.ServiceInstances) {
+		// no changes detected - nothing to update
+		return nil
+	}
+
+	// Update proxy node's service instances
+	proxy.ServiceInstances = instances
```

Comment:
(on my phone so based on memory) in pushConnection do we already do SetServiceInstances? if so we can remove this and get around the race issue?

Ideally all proxy modifications are done in pushConnection do we only have one thread modifying it and avoid races